pkg/ipmasq/ipmasq.go

```diff
@@ -0,0 +1,180 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipmasq
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/ipmasq"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ipmasq")
+)
+
+// ipnet is a wrapper type for net.IPNet to enable de-serialization of CIDRs
+type Ipnet net.IPNet
+
+func (c *Ipnet) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var str string
+	if err := unmarshal(&str); err != nil {
+		return err
+	}
+
+	ip, n, err := net.ParseCIDR(strings.Trim(str, `"`))
+	if err != nil {
+		return fmt.Errorf("Invalid CIDR %s: %s", str, err)
+	}
+	if ip.To4() == nil {
+		return fmt.Errorf("Invalid CIDR %s: only IPv4 is supported", str)
+	}
+
+	*c = Ipnet(*n)
+	return nil
+
+}
+
+// config represents the ip-masq-agent configuration file encoded as YAML
+type config struct {
+	NonMasqCIDRs []Ipnet `yaml:"nonMasqueradeCIDRs"`
+}
+
+// IPMasqMap is an interface describing methods for manipulating an ipmasq map
+type IPMasqMap interface {
+	Update(cidr net.IPNet) error
+	Delete(cidr net.IPNet) error
+	Dump() ([]net.IPNet, error)
+}
+
+// IPMasqAgent represents a state of the ip-masq-agent
+type IPMasqAgent struct {
+	configPath             string
+	nonMasqCIDRsFromConfig map[string]net.IPNet
+	nonMasqCIDRsInMap      map[string]net.IPNet
+	ipMasqMap              IPMasqMap
+}
+
+// Start starts the "ip-masq-agent" controller which is used to sync the ipmasq
+// BPF maps.
+func Start(configPath string, syncPeriod time.Duration) error {
+	return start(configPath, syncPeriod, &ipmasq.IPMasqBPFMap{}, controller.NewManager())
+}
+
+func start(configPath string, syncPeriod time.Duration,
+	ipMasqMap IPMasqMap, manager *controller.Manager) error {
+
+	a := &IPMasqAgent{
+		configPath:        configPath,
+		nonMasqCIDRsInMap: map[string]net.IPNet{},
+		ipMasqMap:         ipMasqMap,
+	}
+
+	if err := a.restore(); err != nil {
+		return err
+	}
+
+	manager.UpdateController("ip-masq-agent",
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) error {
+				return a.Update()
+			},
+			RunInterval: syncPeriod,
+		},
+	)
+
+	return nil
+}
+
+// Update updates the ipmasq BPF map entries with ones from the config file.
+func (a *IPMasqAgent) Update() error {
+	if err := a.readConfig(); err != nil {
+		return err
+	}
+
+	for cidrStr, cidr := range a.nonMasqCIDRsInMap {
+		if _, ok := a.nonMasqCIDRsFromConfig[cidrStr]; !ok {
+			log.WithField(logfields.CIDR, cidrStr).Info("Removing CIDR")
+			a.ipMasqMap.Delete(cidr)
+			delete(a.nonMasqCIDRsInMap, cidrStr)
+		}
+	}
+
+	for cidrStr, cidr := range a.nonMasqCIDRsFromConfig {
```

Comment:
Oh, one more point here. Did you consider the ordering between add/delete here?

Recent similar BPF map updates changes in policy side have shown that the ordering can impact datapath availability, see https://github.com/cilium/cilium/pull/10936

Reply:
I changed the order so that adds happen before deletes just to make new CIDRs available a bit quicker.

The ordering won't have any effect on established connections. However, if there is a connection which previously was masqueraded, and after the CIDR update it's no longer a subject to masquerading, then the egressing packets from the pod after the update won't be masqueraded, and thus it will be dropped. If it's an issue, we can always add a lookup in the BPF (S)NAT map when deciding whether the masquerading is needed. But that would come with a performance penalty in the fastpath.